### PR TITLE
[CARBONDATA-861] Improvements in query

### PR DESCRIPTION
**Pull request by ravipesala**
Following are the list of improvements done in this part of PR.
1. Removed multiple creation of array and copy of it in Dimension and measure chunk readers.
2. Simplified logic of finding offsets of nodictionary keys in the class SafeVariableLengthDimensionDataChunkStore.
3. Avoided byte array creation and copy for nodictionary columns in case of vectorized reader. Instead directly sending the length and offset to vector.
4. Removed unnecessary decoder plan additions to oprtimized plan. It can optimize the codegen flow.
5. Updated CompareTest to take table blocksize and kept as 32 Mb in order to make use of small sorting when doing take ordered in spark.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1380/

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala [hunk 1]
```diff
@@ -306,6 +307,8 @@ object CompareTest {
     // do GC and sleep for some time before running next table
     System.gc()
     Thread.sleep(1000)
+    System.gc()
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: Is this required?

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDataFrameWriter.scala [hunk 2]
```diff
@@ -171,11 +171,20 @@ class CarbonDataFrameWriter(sqlContext: SQLContext, val dataFrame: DataFrame) {
...
+      }
+    ).append(
+      if (options.tableBlockSize.isDefined) {
+        s"'table_blocksize' = '${options.tableBlockSize.get}'"
       } else {
         ""
       }
     )
+    if (property.nonEmpty && property.charAt(property.length-1) == ',') {
+      property = property.replace(property.length-1, property.length, "")
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDataFrameWriter.scala]: change `property.length-1` to `property.length - 1`

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala (hunk 1 above)
Comment (ravipesala) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: There is no guarntee that GC will be called after calling of System.gc(), thats why after waiting for 1 second called again to increase the probability of running GC

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDataFrameWriter.scala (hunk 2 above)
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDataFrameWriter.scala]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1461/

**Comment by jackylk** [comment]
@ravipesala please rebase

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1560/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**